utils/throttler.go

```diff
@@ -16,6 +16,14 @@ type Throttler[T any] struct {
 	maxQueueLen int32
 }
 
+// Question regarding the whole architecture, why buffered channel with limited number
+// of workers is not used? resource busy can be implemented with
+//
+//	select {
+//	  case <- dataChannel: ...
+//	  default:
+//	     err = ErrResourceBusy
+//	 }
```

Comment:
that might work, but you would need 2 channels if I am not mistaken. One to act as a semaphore on the resource and one to act as a queue if semaphore is not available. Sounds a bit tricky to get it right without any race conditions.